### PR TITLE
osd: Reduce backfill/recovery default limits for mClock and other optimizations

### PR DESCRIPTION
**Pull request by sseshasa**
The following summarizes the fixes and optimizations made:

1. Reduce the default number of max backfills/recoveries for mclock to more
    realistic levels for both HDD and SSD.
    (Max Backfills = 10, Max Recoveries_hdd = 10, Max Recoveries_ssd = 20)

2.  Ignore osd bench results for mclock if it exceeds the set threshold. The
     OSD's capacity is retained at the default setting to prevent issues with
     performance. A cluster log warning is raised to indicate this for the osd.

3. Optimize the HDD cost per io and cost per byte parameters. This helps
     in realizing slightly faster recoveries while maintaining similar average
     client throughput.

4. **TBD**: Optimize the mclock profiles related to client, recovery and
    other background ops  allocations. The commit related to this change is
    omitted for now since the new allocations for best effort client ops (e.g.
    scrubs, snaptrim, pg deletes)  was resulting in  very slow scrub progress.
    Further tests are  planned after which the commit will be re-introduced.

In addition to this, PR https://github.com/ceph/ceph/pull/48138 also adds an
optimization to osd bench to perform more randomized writes in order to
get more accurate results for the osd's IOPS performance.

This was tested on a larger scaled cluster and the results are presented further below.

Fixes: https://tracker.ceph.com/issues/57529
Signed-off-by: Sridhar Seshasayee <sseshasa@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by sseshasa** [comment]
Results with the fixes so far look good with the client ops showing comparable or better read and write throughput when compared to WPQ scheduler. The only optimization that didn't help much was the osd bench fix (PR: https://github.com/ceph/ceph/pull/48138) as the issue is probably related to the hardware/controller/disk itself that is resulting in inflated osd bench results. But the fallback mechanism helped in the case and the test proceeded as expected. Results with charts will be shared once all the tests are completed.

Other Changes TBD:
1. Fix standalone tests that check for the backfill limits (ceph-helpers.sh).
2. Documentation of the new profile allocations.
3. Documentation of the new config options.

In the meantime, review of the existing changes can proceed.

**Comment by sseshasa** [comment]
> Other Changes TBD:
> 
>     1. Fix standalone tests that check for the backfill limits (ceph-helpers.sh).
> 
>     2. Documentation of the new profile allocations.
> 
>     3. Documentation of the new config options.
> 
**Update:**
In addition to the above improvements, the current HDD mclock cost parameters `osd_mclock_cost_per_byte_usec_hdd` and `osd_mclock_cost_per_io_usec_hdd` are optimized further to improve the recovery rate while keeping the client ops performance at similar levels. Tests with the optimized HDD cost parameters are showing good results so far. Additional tests are in-progress. If the results continue looking, a new commit with the optimized cost params will be included as well.

**Comment by amathuria** [comment]
jenkins test make check

**Comment by amathuria** [comment]
@neha-ojha this looks good to me.

**Review by vumrao — APPROVED** [review]
We did a lot of OSD failure testing at the Scale cluster starting with 192 OSDs where we found this issue and then we were able to reproduce it in a small 30 OSDs cluster. We tested finally this complete PR in a new Scale 312 OSDs cluster and @sseshasa will provide a test result summary.  After all, this scale testing changes LGTM. Thank you, team!

**Comment by vumrao** [comment]
@sseshasa I think we should open a new tracker for the OSD bench improvement and I think we could use the logic of using FIO to do the OSD bench. I was talking to @neha-ojha. I think we could think of packaging FIO in our Ceph container images so we should be good! Can you please open the tracker, other looks good?

Thank you again for your quick fixes it really helped!

**Comment by sseshasa** [comment]
> We did a lot of OSD failure testing at the Scale cluster starting with 192 OSDs where we found this issue and then we were able to reproduce it in a small 30 OSDs cluster. We tested finally this complete PR in a new Scale 312 OSDs cluster and @sseshasa will provide a test result summary. After all, this scale testing changes LGTM. Thank you, team!

As mentioned by @vumrao, here are the results of the tests that his team ran on the Scale 312 OSDs cluster:

The table captures the client IOPS measured during various phases of the test. The tests involve failing nodes in phases that host the OSDs and measuring the client ops during that phase and also during the recovery phase when PGs are recovering from the node failures:

**COSBench Workoad Ratios:** 60% read,16% write, 10% list, 14% delete

| Test Performed |  read throughput (MB/s) |  % diff vs wpq |
| -------------------- |  ------------------------------- | ------------------- |
| hybrid noFailure | 309(wpq) vs 310(mclock) | Baseline |
| hybrid OSDnode1 Failure | 164(wpq) vs 282(mclock) |  72% improvement with mclock |
| hybrid OSDnode2 Failure | 191(wpq) vs 273(mclock) |  43% improvement with mclock |
| hybrid OSD recover | 230(wpq) vs 282(mclock) |  23% improvement with mclock |

| Test Performed |  write throughput (MB/s) |  % diff vs wpq |
| -------------------- |  ------------------------------- | ------------------ |
| hybrid noFailure | 148(wpq) vs 146(mclock) | Baseline |
| hybrid OSDnode1 Failure | 72(wpq) vs 131(mclock) | 82% improvement with mclock |
| hybrid OSDnode2 Failure | 88(wpq) vs 127(mclock) | 44% improvement with mclock |
| hybrid OSD recover | 110(wpq) vs 132(mclock) | 20% improvement with mclock |

File: doc/rados/configuration/mclock-config-ref.rst [hunk 1]
```diff
@@ -386,6 +411,70 @@ The individual QoS-related config options for the *custom* profile can also be
...
+
+.. important:: The max backfill/recovery options that can be modified are listed
+   in section `Recovery/Backfill Options`_. The modification of the mClock
+   default backfills/recovery limit is gated by the
+   :confval:`osd_mclock_override_recovery_settings` option, which is set to
+   *false* by default. Attempting to modify any default recovery/backfill
+   limits without setting the gating option will reset that option back to the
+   mClock defaults along with a warning message logged in the cluster log. Note
+   that it may take a few seconds for the default value to come back into
+   effect. Verify the limit using the *config get* command as shown below.
```
Comment (neha-ojha) [inline doc/rados/configuration/mclock-config-ref.rst:434]: config show

File: doc/rados/configuration/mclock-config-ref.rst [hunk 2]
```diff
@@ -413,6 +502,40 @@ node whose underlying device type is SSD:
...
+used:
+
+- :confval:`osd_mclock_iops_capacity_threshold_hdd` = 500
+- :confval:`osd_mclock_iops_capacity_threshold_ssd` = 80000
+
+The following automated step is performed:
+
+Fallback to using default OSD capacity (automated)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+If the OSD bench reports a measurement that exceeds the above threshold
```
Comment (neha-ojha) [inline doc/rados/configuration/mclock-config-ref.rst:520]: nit: If OSD bench

File: doc/rados/configuration/mclock-config-ref.rst [hunk 3]
```diff
@@ -413,6 +502,40 @@ node whose underlying device type is SSD:
...
+values depending on the underlying device type, the fallback mechanism reverts
+to the default value of :confval:`osd_mclock_max_capacity_iops_hdd` or
+:confval:`osd_mclock_max_capacity_iops_ssd`. Additionally, a warning is
+logged in the cluster logs indicating the same. The threshold config options
```
Comment (neha-ojha) [inline doc/rados/configuration/mclock-config-ref.rst:524]: You can add an example of the cluster log message.

File: doc/rados/configuration/mclock-config-ref.rst [hunk 4]
```diff
@@ -88,6 +88,14 @@ Users can choose between the following built-in profile types:
 .. note:: The values mentioned in the tables below represent the percentage
           of the total IOPS capacity of the OSD allocated for the service type.
 
+By default, the *high_client_ops* profile is enabled to ensure that a larger
+chunk of the bandwidth allocation goes to client ops. Background recovery ops
+are given lower allocation (and therefore take a longer time to complete). But
+there might be instances that necessitate giving higher allocations to either
+client ops or recovery ops. In order to deal with such a situation, the
+alternate built-in profiles may be enabled by following the steps mentioned
+in the next sections.
```
Comment (neha-ojha) [inline doc/rados/configuration/mclock-config-ref.rst:97]: nit: in next sections

File: doc/rados/configuration/mclock-config-ref.rst [hunk 5]
```diff
@@ -413,6 +502,40 @@ node whose underlying device type is SSD:
...
+following additional step is recommended to address this:
+
+Run custom drive benchmark if defaults are not accurate (manual)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+If the default OSD capacity is not accurate, the recommendation is to run a
+custom benchmark using your preferred tool (e.g. Fio) on the drive and then
+override the ``osd_mclock_max_capacity_iops_[hdd, ssd]`` option as described
+in the `Specifying  Max OSD Capacity`_ section.
+
+This step is highly recommended until an alternate automated mechansim is worked
```
Comment (neha-ojha) [inline doc/rados/configuration/mclock-config-ref.rst:537]: typo: mechanism

**Review by neha-ojha — APPROVED** [review]
LGTM!

**Comment by neha-ojha** [comment]
jenkins test make check

File: doc/rados/configuration/mclock-config-ref.rst (hunk 1 above)
Comment (sseshasa) [reply doc/rados/configuration/mclock-config-ref.rst]: Fixed.

File: doc/rados/configuration/mclock-config-ref.rst (hunk 2 above)
Comment (sseshasa) [reply doc/rados/configuration/mclock-config-ref.rst]: Fixed.

File: doc/rados/configuration/mclock-config-ref.rst (hunk 3 above)
Comment (sseshasa) [reply doc/rados/configuration/mclock-config-ref.rst]: Added the cluster log example.

File: doc/rados/configuration/mclock-config-ref.rst (hunk 4 above)
Comment (sseshasa) [reply doc/rados/configuration/mclock-config-ref.rst]: Fixed.

File: doc/rados/configuration/mclock-config-ref.rst (hunk 5 above)
Comment (sseshasa) [reply doc/rados/configuration/mclock-config-ref.rst]: Fixed.

**Comment by sseshasa** [comment]
@neha-ojha @amathuria Test result is updated after removing the commit related to profile allocations. All the scrub tests passed. For a few erasure-code, osd and osd-backfill related standalone tests, a subset of the tests were executed again after settting the _osd_mclock_override_recovery_settings_ flag to _true_ since they modify the recovery/backfill limits. Please take a look. Thanks!

**Comment by neha-ojha** [comment]
jenkins test dashboard

**Pull request merged**